include/fiction/algorithms/simulation/sidb/energy_distribution.hpp

```diff
@@ -8,76 +8,215 @@
 #include "fiction/technology/charge_distribution_surface.hpp"
 #include "fiction/technology/constants.hpp"
 
+#include <algorithm>
 #include <cmath>
+#include <cstddef>
 #include <cstdint>
+#include <cstdlib>
+#include <iterator>
 #include <map>
+#include <optional>
 #include <set>
+#include <utility>
 #include <vector>
 
 namespace fiction
 {
 
 /**
- * Data type to collect electrostatic potential energies (unit: eV) of charge distributions with corresponding
- * degeneracy (i.e., how often a certain energy value occurs).
+ * This class is used to store the energy distribution of an SiDB layout. The energy distribution is a map that contains
+ * the electrostatic potential as a key and its degeneracy as a value. To be more precise, if two different charge
+ * distributions occur with the same energy, the degeneracy value of the energy state is two.
  */
-using sidb_energy_distribution = std::map<double, uint64_t>;  // unit: (eV, unit-less)
+class sidb_energy_distribution
+{
+  public:
+    sidb_energy_distribution() = default;
+    /**
+     * Returns the nth state (energy + degeneracy) in the energy distribution.
+     *
+     * @param state_index The index of the state to be retrieved.
+     * @return The energy and degeneracy of the state at the specified index. If the index is out of range,
+     * `std::nullopt` is returned instead.
+     */
+    [[nodiscard]] std::optional<std::pair<double, uint64_t>> get_nth_state(const uint64_t state_index) const noexcept
```

Comment:
Let's factor out this return type as its own type, e.g., `energy_state`.

Reply:
good idea.